txtorcon/torconfig.py

```diff
@@ -1129,13 +1129,24 @@ def save(self):
             return defer.succeed(self)
 
         args = []
+        directories = []
+        hidden_services = []
         for (key, value) in self.unsaved.items():
             if key == 'HiddenServices':
-                self.config['HiddenServices'] = value
                 for hs in value:
                     for (k, v) in hs.config_attributes():
-                        args.append(k)
-                        args.append(v)
+                        if k == 'HiddenServiceDir':
+                            if v not in directories:
+                                directories.append(v)
+                                hidden_services.append(hs)
+                                args.append(k)
+                                args.append(v)
+                            else:
+                                raise RuntimeError("Trying to add hidden service with same HiddenServiceDir: %s" % v)
+                        else:
+                            args.append(k)
+                            args.append(v)
+                self.config['HiddenServices'] = hidden_services
```

Comment:
Does this get properly wrapped in a ListWrapper_  when set this way?

Also, I'm not sure the point here: you're saving up the "new" hidden-services list, but if there was ever a redundant HS-dir this line would never run anyway, because we'd error-out...

I think take out the "re-make the list" code in here: we're basically saying "it's an error if you have redundant hs-dirs, and so you can't save" and then it's up to the user to fix that.